File: 200.py
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        m = len(grid)
        n = len(grid[0])

        if m == 0 or n == 0:
            return 0

        def DFS(i, j):
            grid[i][j] = 0

            if i + 1 < m and grid[i + 1][j] == '1':
                DFS(i + 1, j)
            if i > 0 and grid[i - 1][j] == '1':
                DFS(i - 1, j)
            if j + 1 < n and grid[i][j + 1] == '1':
                DFS(i, j + 1)
            if j > 0 and grid[i][j - 1] == '1':
                DFS(i, j - 1)
        
        islandCount = 0
        for i in range(m):
            for j in range(n):
                if grid[i][j] == '1':
                    islandCount += 1
                    DFS(i, j)
        
        return islandCount

File: 200.cpp
class Solution {
public:
    void dfs(int i, int j, vector<vector<char>>& grid, int n, int m) {
        if (!(i >= 0 && i < n && j >= 0 && j < m)) {
            return;
        }

        if (grid[i][j] == '1') {
            grid[i][j] = '*'; // marking as visited

            dfs(i + 1, j, grid, n, m);
            dfs(i, j + 1, grid, n, m);
            dfs(i - 1, j, grid, n, m);
            dfs(i, j - 1, grid, n, m);
        }
    }

    int numIslands(vector<vector<char>>& grid) {

        // Magically Reducing Run Time
        ios_base::sync_with_stdio(false);
        cin.tie(NULL);
        cout.tie(NULL);

        int n = grid.size();
        int m = grid[0].size();

        int num_islands = 0;

        for(int i = 0; i < n; i++) {
            for(int j = 0; j < m; j++) {
                if (grid[i][j] == '1') {
                    num_islands++;
                    dfs(i, j, grid, n, m);
                }
            }
        }

        return num_islands;

    }
};